astropy/table/index.py

```diff
@@ -0,0 +1,630 @@
+        index = super(Index, Index).__new__(Index)
+        index.__init__(None, impl=self.data.__class__, col_dtypes=
+                      [x.info.dtype for x in self.columns])
+        index.data = deepcopy(self.data, memo)
+        index.columns = self.columns[:] # new list, same columns
+        memo[id(self)] = index
+        return index
+
+class SlicedIndex(object):
+    '''
+    This class provides a wrapper around an actual Index object
+    to make index slicing function correctly. Since numpy expects
+    array slices to provide an actual data view, a SlicedIndex should
+    retrieve data directly from the original index and then adapt
+    it to the sliced coordinate system as appropriate.
+
+    Parameters
+    ----------
+    index : Index
+        The original Index reference
+    index_slice : slice
+        The slice to which this SlicedIndex corresponds
+    original : bool
+        Whether this SlicedIndex represents the original index itself.
+        For the most part this is similar to index[:] but certain
+        copying operations are avoided, and the slice retains the
+        length of the actual index despite modification.
+    '''
+    def __init__(self, index, index_slice, original=False):
+        self.index = index
+        self.original = original
+        self._frozen = False
+
+        if isinstance(index_slice, tuple):
+            self.start, self._stop, self.step = index_slice
+        else: # index_slice is an actual slice
+            num_rows = len(index.columns[0])
+            self.start, self._stop, self.step = index_slice.indices(num_rows)
+
+    @property
+    def length(self):
+        return 1 + (self.stop - self.start - 1) // self.step
+
+    @property
+    def stop(self):
+        '''
+        The stopping position of the slice, or the end of the
+        index if this is an original slice.
+        '''
+        return len(self.index) if self.original else self._stop
+
+    def __getitem__(self, item):
+        '''
+        Returns another slice of this Index slice.
+
+        Parameters
+        ----------
+        item : slice
+            Index slice
+        '''
+        if self.length <= 0:
+            # empty slice
+            return SlicedIndex(self.index, slice(1, 0))
+        start, stop, step = item.indices(self.length)
+        new_start = self.orig_coords(start)
+        new_stop = self.orig_coords(stop)
+        new_step = self.step * step
+        return SlicedIndex(self.index, (new_start, new_stop, new_step))
+
+    def sliced_coords(self, rows):
+        '''
+        Convert the input rows to the sliced coordinate system.
+
+        Parameters
+        ----------
+        rows : list
+            Rows in the original coordinate system
+
+        Returns
+        -------
+        sliced_rows : list
+            Rows in the sliced coordinate system
+        '''
+        if self.original:
+            return rows
+        elif self.step > 0:
```

Comment:
It should be possible to do this with numpy array operations, e.g.:
```
row0 = row - self.start
if self.step != 1:
    ok = np.mod(row0, self.step) == 0
    row0 = row0[ok]
ok = (self.start <= row0) & (row0 < self.stop)
return row0[ok] // self.step
```
In general you don't want any pure-Python loops or list comprehensions where you could be looping over anything of order the table length (which could be 10^9 or more).